### ISSUE TITLE
[dev-log] UGB

### ISSUE THREAD

**Issue by dkluffy** [dkluffy/dklog#5]
# UGB - dev log

## TimeLine

* 2020/02/02 不使用mobilenet，决定重新写一个，先测试gan
    + [GAN，类似问题](https://blog.csdn.net/think_three/article/details/88714237)
    + [2-channel，可以参考](https://blog.csdn.net/hjimce/article/details/50098483)

* 2020/02/02 测试了softmax
    准确率大概只有0.6
* 2020/01/31 已经实现了大部分训练功能
    + tensorboard
    + save_weights,load_model,checkpoints，scheduler，。。。
    + 用dataset大大简化了数据生成问题
    + 训练了50 epoch,loss change from 2.1244 to 1.0398

```python
100/100 [==============================] - 86s 861ms/step - loss: 1.0488
Epoch 45/50
100/100 [==============================] - 86s 858ms/step - loss: 1.0475
Epoch 46/50
100/100 [==============================] - 90s 905ms/step - loss: 1.0451
Epoch 47/50
100/100 [==============================] - 81s 815ms/step - loss: 1.0443
Epoch 48/50
100/100 [==============================] - 86s 860ms/step - loss: 1.0425
Epoch 49/50
100/100 [==============================] - 85s 855ms/step - loss: 1.0412
Epoch 50/50
100/100 [==============================] - 86s 861ms/step - loss: 1.0398

"""
loss可能卡在1附近不动了， 用较大的LR和极小LR测试，LOSS都没能成功下来，可能还要调整数据集
暂时先放一下，还有别事情要做
"""
#TODO:
# 调整lr 和 optimizer，batch_size等
# 调整数据集
# 调整model,从目前训练来看，训练过的weights，貌似没有帮助
# 实现metric


```

## 目标

- 高精度（不受分辨率、大小影响）

- 不需要直接返回坐标，不需要端到端

- 模型可以简单训练

- oneshot:旧版利用opencv实现的几乎时zero-shot，但受限于精度

- 最少的人工标签

### 分块对比法-tripletloss版

通过把X（截屏得到的图片）裁剪成N个固定大小(128,128,3)小块$x_i$，对比$x_i$和模板$T_i$ 的L2 距离。

#### 训练方法

- 双生网络，人脸认证的思想，对比$x_i,t_i$的encode l2距离
- from tensorflow_addons.losses import TripletSemiHardLoss
- 使用训练过的mobilenetv2权值
- 用noise和模板的部分截图$t_i$（模拟实际$x_i$）训练模型

#### 验证

- 本地验证(todo)
- 生成vector和meta，导入https://projector.tensorflow.org/ (UMAP)

**----------**

### 分块对比法-classification

**----------**

### 分块对比法-GAN

用GAN的方法，可以同时得到G和D，值得一试

**----------**

### FCN

SPP/FASTER R-CNN/YOLOV3
这个时我最不想做的，因为需要带坐标的数据

**Comment by dkluffy** [comment]
- 破显卡在WINDOWS下可用内存1G都没有
- 在COLAB上训练了150次，还没有测试，训练过程的验证mse大多在0.1以下，偶尔有几个批次特别高

**Comment by dkluffy** [comment]
=.= 崩溃了
- 先是发现，模型输出有问题，改了输出
- 把GAN的模型拆开来，只训练识别部分的网络，发现我的模型根本没用，LOSS完全没变化
- 模型连两张相同图片都不能识别

**Comment by dkluffy** [comment]
有切换到SOFTMAX测试， 不是我想要的
好像还是一样的问题，模型的训练LOSS不下降，虽然ACC有上去

应该是欠拟合，LOSS没有下降，说明模型没有收敛，没有学到特征

不搞了，明天再说

**Comment by dkluffy** [comment]
> 有切换到SOFTMAX测试， 不是我想要的
> 好像还是一样的问题，模型的训练LOSS不下降，虽然ACC有上去
> 
> 应该是欠拟合，LOSS没有下降，说明模型没有收敛，没有学到特征
> 
> 不搞了，明天再说

怎么解决？降低LR貌似也没用，难道要对输入做STD？

**Comment by dkluffy** [comment]
今天测试了下用INCEPTION模型， 没有重用WEIGHTS从头开始训练，接的SOFTMAX(64),用1e-4的LR训练了50个EPOCH后，
ACC大概有90%+ ， 但是LOSS始终没有突破到3以下，应该是 `数据不够`

**Comment by dkluffy** [comment]
今天产生了V7， 又回到MOBILENETV2， 直接加一个SOFTMAX(16,只有16个图)
结果感人，和INCEPTION类似结果。

**Comment by dkluffy** [comment]
![image](https://user-images.githubusercontent.com/6056493/74837767-31c0bc00-535d-11ea-8d3d-ab23287fc535.png)

这个是MOBILENETV2的55 *200STEPS训练结果

EPOCH-[55/100]-  DISC LOSS:1.8746 ,val_loss:1.9574,acc:0.8636 , (200/200)

**Comment by dkluffy** [comment]
> ![image](https://user-images.githubusercontent.com/6056493/74837767-31c0bc00-535d-11ea-8d3d-ab23287fc535.png)
> 
> 这个是MOBILENETV2的55 *200STEPS训练结果
> 
> EPOCH-[55/100]- DISC LOSS:1.8746 ,val_loss:1.9574,acc:0.8636 , (200/200)

结果还行吧，但是，还不足以用来替代opencv的template match

**Comment by dkluffy** [comment]
今天 不甘心，又重写了一遍GAN， 数学基础不行，没法证明LOSS是否有效，模型是否有效，只能通过实验来证明。

### 在这件事情上花太多时间了，好多其他的东西都没学没练，感觉马上就要忘记光了 =。=

**Comment by dkluffy** [comment]
GAN 训练结果：（v8)
100*100*4次迭代后， 完全没用。 discriminator 会把奇怪的图片认为是匹配的，而只有少数亮度特别高的识别为不同。

生成器也还没能生成有用的图片。

=.= 太难了，没有计算资源的情况下，还是别研究深度卷积网络了 =。=

**Comment by dkluffy** [comment]
有V9的构思了，找到工作前，禁止做 =。=
暂时关闭这个LOG

**Comment by dkluffy** [comment]
v11 - lossless triplet，
 训练结果和softmax接近， 用了一个浅网络，训练了50*100 次的效果

训练比之前的版本都快一点， 再训练一会看看

![image](https://user-images.githubusercontent.com/6056493/75036480-ec32f900-54ec-11ea-82a1-dd83cb8ed210.png)

**Comment by dkluffy** [comment]
=。= 真的不做了， 最好的方案，还是YOLO+SSP这样的方案是最好的。
其他方案 虽然不需要大量标签，但是不见得 比YOLO 容易训练。

以下这个模板匹配的论文， 训练也需要大量标签，还不如直接用YOLO
[817.pdf](https://github.com/dkluffy/dklog/files/4236269/817.pdf)
